// testcafe/lib/browser/provider/built-in/dedicated/chrome/requests-interceptor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const testcafe_hammerhead_1 = require("testcafe-hammerhead");
const connection_1 = __importDefault(require("../../../../connection"));
const injectables_1 = require("../../../../../assets/injectables");
const HTTP_STATUS_OK = 200;
class RequestsInterceptor {
    constructor(browserId) {
        this._browserId = browserId;
    }
    _getResponseAsString(response) {
        return response.base64Encoded
            ? Buffer.from(response.body, 'base64').toString()
            : response.body;
    }
    async _prepareInjectableResources() {
        const browserConnection = connection_1.default.getById(this._browserId);
        const proxy = browserConnection.browserConnectionGateway.proxy;
        const windowId = browserConnection.activeWindowId;
        const taskScript = await browserConnection.currentJob.currentTestRun.session.getTaskScript({
            referer: '',
            cookieUrl: '',
            isIframe: false,
            withPayload: true,
            serverInfo: proxy.server1Info,
            windowId,
        });
        const injectableResources = {
            stylesheets: [
                injectables_1.TESTCAFE_UI_STYLES,
            ],
            scripts: [
                ...testcafe_hammerhead_1.INJECTABLE_SCRIPTS,
                ...injectables_1.SCRIPTS,
            ],
            embeddedScripts: [taskScript],
        };
        injectableResources.scripts = injectableResources.scripts.map(script => proxy.resolveRelativeServiceUrl(script));
        injectableResources.stylesheets = injectableResources.stylesheets.map(style => proxy.resolveRelativeServiceUrl(style));
        return injectableResources;
    }
    async setup(client) {
        const fetchAllDocumentsPattern = {
            urlPattern: '*',
            resourceType: 'Document',
            requestStage: 'Response',
        };
        await client.Fetch.enable({ patterns: [fetchAllDocumentsPattern] });
        client.Fetch.on('requestPaused', async (params) => {
            const { requestId, responseHeaders, responseStatusCode, } = params;
            const responseObj = await client.Fetch.getResponseBody({ requestId });
            const responseStr = this._getResponseAsString(responseObj);
            const injectableResources = await this._prepareInjectableResources();
            const updatedResponseStr = testcafe_hammerhead_1.injectResources(responseStr, injectableResources);
            await client.Fetch.fulfillRequest({
                requestId,
                responseCode: responseStatusCode || HTTP_STATUS_OK,
                responseHeaders: responseHeaders || [],
                body: Buffer.from(updatedResponseStr).toString('base64'),
            });
        });
    }
}
exports.default = RequestsInterceptor;
module.exports = exports.default;
